### ISSUE TITLE
SqsSubscription: true fails to deploy

### ISSUE THREAD

**Issue by jlhood** [aws/serverless-application-model#1424]
**Description:**

Found a bug while using SAM's new SQS queue automatic generation with the SNS Event Type feature: If your Lambda function timeout is greater than 30 seconds and you're using the `SqsSubscription: true` property to create an SQS queue with default settings, your deployment will fail with an error from the Lambda service saying the Lambda function timeout cannot be greater than the SQS queue's visibility timeout. SAM should pass the Lambda function timeout through as the SQS queue visibility timeout.

Sample template:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Resources:
  Topic:
    Type: AWS::SNS::Topic

  MyFunction:
    Type: AWS::Serverless::Function
    Properties:
      InlineCode: |
        exports.handler = async (event) => {
          return {}
        }
      Handler: index.handler
      Runtime: nodejs12.x
      Timeout: 60
      Events:
        TopicEvent:
          Type: SNS
          Properties:
            Topic: !Ref Topic
            SqsSubscription: true
```

**Steps to reproduce the issue:**

1. Attempt to deploy the above template.

**Observed result:**

Deployment fails with this error when trying to create the Lambda event source mapping:

> Queue visibility timeout: 30 seconds is less than Function timeout: 60 seconds (Service: AWSLambda; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: ****) The following resource(s) failed to create: [MyFunctionTopicEventEventSourceMapping]. . Rollback requested by user.

**Expected result:**

Successful deployment.

**Comment by piersf** [comment]
I got hit by this as well a couple of times today. It appears that you are forced to set the visibility timeout >= to the function's timeout. Although most of the time this should be the case(queue visibility timeout should be at least 6 times greater the Lambda timeout value), there are cases where you need to set the visibility time out to a low value since this controls implicitly the retry interval of the Lambda function. 

We have a use-case where the Producer sends a message(s) every 5 minutes, and it will keep sending that same message until it has been processed and confirmed by the Consumer(single). Now if the Consumer fails the first time to process that message(for whatever reason), we want it to retry as fast as possible(for up to X times) and before the deduplication interval expires. 
But since the Lambda retry interval is implicitly controlled by the queue's visibility timeout, then if this timeout is much greater than the Lambda timeout, the Lambda retry will happen way after the deduplication interval(5 min) has expired. This will result in duplicates in the queue(FIFO).

In our case, we set the queue visibility timeout to a lower value than the Lambda timeout because of the 5 minutes interval that the Producer sends messages. It's a trade-off between giving enough time to the Consumer but also waiting for as little as possible for the same message to be available again for processing, in case of a failure the first time. Otherwise, with a higher visibility timeout, it takes more time for the same message to be reprocessed. And by that time, the deduplication interval(5 min) might have expired and we risk having duplicates in the FIFO queue as the same message will be sent again after 5 minutes when the Producer runs since it won't have been processed/confirmed by the Consumer yet.

Is the above issue really a SAM bug or is this the doing of CloudFormation behind the scenes? Is there a solution to the above-described issue?

Thanks

**Comment by ShreyaGangishetty** [comment]
@piersf SAM needs to add `VisibilityTimeout` property for SQS queue resource that SAM generates and set the value to `Timeout` given for the Lambda function. The value of `VisibilityTimeout` being always greater than or equal to Lambda `Timeout` is not set by SAM.

**Comment by pavelrvia** [comment]
IMO this constraint shouldn't exist. This should be a warning, not an error. If I want to set the queue visibility timeout to be less than the lambda timeout, because of some weird use case that I have, then I should be able to do that. It isn't CF's job to enforce this.

**Comment by CCBet** [comment]
@praneetap, @ShreyaGangishetty  can you please confirm (or infirm, if it's the case) the following statement? 

For the moment, we cannot set the **default visibility timeout for a SQS queue** to be **smaller** than the **timeout of the Lambda** which is consuming the SQS queue messages.

I'm asking this because my team ran today into a similar CloudFormation change set execution issue: 
_**"Queue visibility timeout: 30 seconds is less than Function timeout: 900 seconds (Service: AWSLambda; Status Code: 400; Error Code: InvalidParameterValueException; ... "**_ 

If this is true, we are forced to have the SQS default visibility timeout = Lambda timeout

Also, a quite misleading (if not actually a bug) thing was the fact that this error was thrown in CloudFormation, but when we changed the SQS default visibility timeout from AWS console, it worked. This means that the **probable** equality constraint is not checked when applying the update from AWS console, but it is enforced in CloudFormation. 
The AWS console update is not a valid choice for us because we deploy our changes from infrastructure code. The AWS console update was made just for testing the update which should be made from the infrastructure code -> CloudFormation.

**Comment by michaelj-smith** [comment]
The purpose of this restriction is to prevent duplicate processing of messages. If your Lambda takes 900 seconds maximum to process a message, and the VisibilityTimeout is 30 seconds, then SQS will invoke your Lambda function every 30 seconds until it gets a confirmed result (after 900 seconds). This restriction is meant as a convenience to you, to help you avoid poor architectural decisions. There are ways to work around it if you are determined.

**Comment by piersf** [comment]
@michaelj-smith what ways are there to work around this?

Since the `VisibilityTimeout` controls implicitly the execution interval of the function, some times there are use-cases where one would want to have the function executed as fast as possible after a previously failed execution, something that would be done by lowering the visibility timeout value. 

But currently, considering that the queue timeout can't be lower than the function timeout, we can't achieve this.

**Comment by ajdnik** [comment]
@michaelj-smith is there a way to set `VisibilityTimeout` in SAM template when using `SqsSubscription: true`? Or does the SQS need to be created ahead of time in that case?

@piersf One workaround is to `sam deploy` with `SqsSubscription: true` and function timeout set to 30. After the deploy succeeds change timeout to your preferred value and call `sam deploy` again. This will keep the `VisibilityTimeout` of SQS at 30 seconds and change Lambda timeout to your preferred value. Of course, as @michaelj-smith explained this is a poor architectural decision.

**Comment by piersf** [comment]
@ajdnik thanks! We did something similar to that. 

Yep, we know the visibility timeout should be at least 5-6 times higher than the function's timeout.

But we have a specific use-case where the Producer sends a message(s) every 5 minutes, and it will keep sending that same message until it has been processed and confirmed by the Consumer(single). Now if the Consumer fails the first time to process that message(for whatever reason), we want it to retry as fast as possible(for up to X times) and before the deduplication interval expires.
But since the Lambda retry interval is implicitly controlled by the queue's visibility timeout, then if this timeout is much greater than the Lambda timeout, the Lambda retry will happen way after the deduplication interval(5 min) has expired. This will result in duplicates in the queue(FIFO)in our scenario since the Producer will keep sending the same message.

**Comment by cailenmusselman** [comment]
> The purpose of this restriction is to prevent duplicate processing of messages. If your Lambda takes 900 seconds maximum to process a message, and the VisibilityTimeout is 30 seconds, then SQS will invoke your Lambda function every 30 seconds until it gets a confirmed result (after 900 seconds).

I get that the goal is to stop developers from shooting themselves in the foot. But the above isn't true if they follow the AWS docs/guidelines: "If you don't know how long it takes to process a message, create a heartbeat for your consumer process" https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html

The length of time it takes to process a message in my SQS varies greatly depending on the payload it contains. I set my lambda timeout to 900s, and my queue's visibility timeout to 60s. But as long as the message is still being processed and hasn't failed I update the visibility timeout for that specific message using the heartbeat approach recommended.

Maybe there could be an additional option added to indicate "I (think) I know what I'm doing".

**Comment by eyalroth** [comment]
Looks like this also affects queues that are explicitly defined and referenced in the lambda:
```yaml
Resources:
  MyLambda:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: MyLambda
      Timeout: 60
      Events:
        Queue:
          Type: SQS
          Properties:
            BatchSize: 1
            Queue: !GetAtt MyQueue.Arn
  MyQueue:
    Type: AWS::SQS::Queue
    Properties:
      QueueName: MyQueue
      VisibilityTimeout: 30
```

What's more, the error seems to only show up when creating a new lambda (or renaming an existing one which basically creates a new one).

When only updating the timeouts of an existing lambda and/or queue -- either increasing the lambda timeout above the queue, or lowering the queue below the lambda -- the error does not seem to show and the stack update succeeds. This is quite an inconsistent behavior.

**Comment by jfuss** [comment]
The constraint/validation is from the AWS Lambda service, which is not anything SAM controls. 

I spoke with the team on the original request. The current thinking is, if you need more control over the SQS that is created to provide the SQS. This can be done without impact by keeping the same name/logical id of the resources SAM creates, if you are in that position.

An example SAM template using `SqsSubscription: true`:
```
Topic:
    Type: AWS::SNS::Topic

MyFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: hello_world/
      Handler: app.lambda_handler
      Runtime: python3.9
      Timeout: 60
      Events:
        TopicEvent:
          Type: SNS
          Properties:
            Topic: !Ref Topic
            SqsSubscription: true
```

Would then be:
```
Topic:
    Type: AWS::SNS::Topic

  MyFunctionTopicEventQueue:
    Type: AWS::SQS::Queue
    Properties:
      VisibilityTimeout: 60

  MyFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: hello_world/
      Handler: app.lambda_handler
      Runtime: python3.9
      Timeout: 60
      Events:
        TopicEvent:
          Type: SNS
          Properties:
            Topic: !Ref Topic
            SqsSubscription:
              QueueArn: 
                Fn::GetAtt: 
                  - MyFunctionTopicEventQueue
                  - "Arn"
              QueueUrl: !Ref MyFunctionTopicEventQueue
```

The LogicalId of the SQS Resource follows the format: `<FunctionLogicalId><EventLogicalId>Queue`.

The reasoning behind this is this allows more control over the SQS resource to you when SAM's defaults aren't right for your use case. Adding `VisibilityTimeout` to the Spec was considered but as we have seen with other generated resources, SAM has to continually expand and keep up with the underlying resource changes.

**Comment by eyalroth** [comment]
@jfuss I don't think I follow your explanation. Is it not possible to add a default `VisibilityTimeout` to queues based on the timeout of the lambda that they're mapped to (all in SAM)?

**Comment by jfuss** [comment]
@eyalroth Not sure what you mean by "all in SAM" as you can do it today just SAM doesn't do it by default. The way to do this is by providing your own `SQS` to the `SqsSubscription` Property.

We cannot add `VisibilityTimeout` by default based on the Lambda `Timeout`. This risk applications that are already on the current default to change underneath them, which we consider a breaking change.

**Comment by eyalroth** [comment]
> @eyalroth Not sure what you mean by "all in SAM" as you can do it today just SAM doesn't do it by default. The way to do this is by providing your own `SQS` to the `SqsSubscription` Property.

By "all in SAM" I meant that SAM -- and not SQS -- would fill in the default.

> We cannot add `VisibilityTimeout` by default based on the Lambda `Timeout`. This risk applications that are already on the current default to change underneath them, which we consider a breaking change.

I see. What about adding a default visibility timeout based on the lambda timeout for cases where the lambda timeout exceeds the current default visibility timeout?

For instance, given that the current default visibility timeout is 30 (is it?), the new default behavior would make it so that if there's a lambda timeout of 30 or more, the default visibility timeout would be resolved to `lambda timeout + 1`.

I'm thinking this won't break anything since currently there is no option to define a queue that doesn't explicitly define the visibility timeout (i.e using the default) and is connected to a lambda with a timeout of 30/default or more, since that would throw an error.

**Comment by jfuss** [comment]
> I see. What about adding a default visibility timeout based on the lambda timeout for cases where the lambda timeout exceeds the current default visibility timeout?

How should we handle when the default changes in the underlying resources?

Typically, SAM tries to avoid validation and assumptions on the underlying resources. This has become brittle when the services change things. Instead, we let as much of the validation happen by CloudFormation or the AWS Service. So in this case, we would need to assume 30 will never change and if it does, we are back in this same conversation.

We did talk about adding `VisibilityTimeout` as a way to configure directly in SAM. What this becomes is SAM trying to keep up with the underlying resources as new configs come up/need configuration. This is why we landed on leaving what we have as is for the "simple" cases and when you need more configuration to provide the SQS.

**Comment by alexagranov7** [comment]
I just encountered this as a CloudFormation error. Queue visibility timeout < lambda timeout causes the error.

**Comment by rickgj** [comment]
Any body fix this error? 
Weird thing is this error only occur in my new SQS Lambda, other SQS Lambdas don't hit.

**Comment by helseidy1046** [comment]
> Any body fix this error? Weird thing is this error only occur in my new SQS Lambda, other SQS Lambdas don't hit.

We had a similar case in which we deployed visibilityTimeout equal to lambda timeout in CDK. After deployment we realized we need to increase the lambda duration and updating it worked fine. Then we ended up with lambda timeout > sqs visibilityTimeout.

I think that was the reason, is that the restriction was somehow applied when creating the queue but not necessarily when updating the lambda. It is still very confusing, but that's the only explanation I can find.
